R/ratio_calculations.R

```diff
@@ -62,3 +66,281 @@ multiple_matrix_ratios <- function(data, matrix_col, conc_col, units_col,
+#'
+#' @param data A PKNCAresults object or its result data.frame.
+#' @param parameter Character. The PPTESTCD value to use for the calculation (e.g., "AUCINF").
+#' @param match_cols Character vector of column names to match between test and reference groups
+#'  or a data.frame specifying columns and values.
+#' @param ref_groups A data.frame specifying reference groups.
+#' At its minimum, contains the contrast variable value(s) for the reference.
+#' @param test_groups A data.frame specifying test groups. Optional.
+#' By default is NULL, allowing rows not in ref_groups be used as test.
+#' @param adjusting_factor Numeric to multiply the ratio. Default is 1.
+#' @param custom_pptestcd Optional character. If provided, will be used as the PPTESTCD value.
+#' @return A data.frame result object with the calculated ratios.
+#' @export
+calculate_ratios <- function(
+  data,
+  parameter,
+  match_cols,
+  ref_groups,
+  test_groups = NULL,
+  adjusting_factor = 1,
+  custom_pptestcd = NULL
+) {
+  UseMethod("calculate_ratios", data)
+}
+#' @export
+calculate_ratios.data.frame <- function(
+  data,
+  parameter,
+  match_cols,
+  ref_groups,
+  test_groups = NULL,
+  adjusting_factor = 1,
+  custom_pptestcd = NULL
+) {
+  if (!any(data$PPTESTCD == parameter)) {
+    warning(
+      paste0(
+        "No parameter with PPTESTCD: '",
+        paste(parameter, collapse = ","),
+        "' is not found in the PKNCA results."
+      )
+    )
+  }
+
+  # Deduce what are all the group columns in the result data.frame
+  extra_res_cols <- c(
+    "PPTEST", "PPORRES", "PPORRESU", "PPSTRES", "PPSTRESU", "type_interval", "exclude"
+  )
+  group_cols <- setdiff(colnames(data), extra_res_cols)
+
+  # Filter for the parameter of interest
+  df <- data[data$PPTESTCD == parameter, ]
+
+  # Define the denominator rows
+  df_den <- merge(df, ref_groups)
+
+  # Define the test rows, which should exclude the ref_groups
+  if (!is.null(test_groups)) {
+    df_num <- merge(df, test_groups)
+  } else {
+    df_num <- df
+    df_num <- anti_join(df_num, df_den, by = intersect(names(df_num), names(df_den)))
+  }
+
+  # Join test and denominator by their matching columns
+  merge(df_num, df_den, by = c(match_cols, "PPTESTCD"), suffixes = c("", "_den")) %>%
+    # If possible compute conversion factors for the units of test and denominator
+    mutate(
+      PPORRESU_factor = get_conversion_factor(PPORRESU_den, PPORRESU),
+      PPSTRESU_factor = if ("PPSTRESU" %in% names(.)) {
+        get_conversion_factor(PPSTRESU_den, PPSTRESU)
+      } else {
+        NULL
+      },
+      PPORRES_den = ifelse(!is.na(PPORRESU_factor), PPORRES_den * PPORRESU_factor, PPORRES_den),
+      PPSTRES_den = if ("PPSTRESU" %in% names(.)) {
+        ifelse(!is.na(PPSTRESU_factor), PPSTRESU_factor * PPSTRES_den, PPSTRES_den)
+      } else {
+        NULL
+      }
+    ) %>%
+    group_by(across(any_of(c(match_cols, group_cols, "PPTESTCD", paste0(group_cols, "_den"))))) %>%
+    unique() %>%
```

Comment:
nitpick: `dplyr` style, I have a habit of using `unique()` with tibbles as well, but should be `distinct` instead:

```suggestion
    distinct() %>%
```